File: android/src/main/java/com/reveila/android/SovereignMemoryManager.java
package com.reveila.android;

import android.content.Context;
import android.util.Log;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.biometric.BiometricManager;
import androidx.biometric.BiometricPrompt;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.FragmentActivity;

import java.util.concurrent.Executor;

/**
 * Java implementation of the Biometric Gatekeeper.
 * Ensures that high-risk actions are protected by hardware-backed authentication.
 */
public class SovereignMemoryManager {

    private static final String TAG = "SovereignMemoryManager";
    private final Context context;

    public SovereignMemoryManager(Context context) {
        this.context = context;
    }

    /**
     * Wrap high-risk writes/actions in a BiometricPrompt challenge.
     * * @param activity   The host activity (The Expo Shell Activity)
     * @param actionName Label for the prompt
     * @param action     The task to execute upon success
     */
    public void secureWriteOperation(FragmentActivity activity, String actionName, Runnable action) {
        Executor executor = ContextCompat.getMainExecutor(activity);

        BiometricPrompt biometricPrompt = new BiometricPrompt(activity, executor,
                new BiometricPrompt.AuthenticationCallback() {
                    @Override
                    public void onAuthenticationError(int errorCode, @NonNull CharSequence errString) {
                        super.onAuthenticationError(errorCode, errString);
                        Log.e(TAG, "Biometric authentication error for " + actionName + ": " + errString);
                        Toast.makeText(activity, "Biometric Error: " + errString, Toast.LENGTH_LONG).show();
                    }

                    @Override
                    public void onAuthenticationSucceeded(@NonNull BiometricPrompt.AuthenticationResult result) {
                        super.onAuthenticationSucceeded(result);
                        Log.i(TAG, "Biometric verified successfully. Executing " + actionName + "...");
                        if (action != null) {
                            action.run(); // Execute the secure action
                        }
                    }

                    @Override
                    public void onAuthenticationFailed() {
                        super.onAuthenticationFailed();
                        Log.w(TAG, "Biometric authentication failed. Action " + actionName + " blocked.");
                        Toast.makeText(activity, "Authentication failed. Try again.", Toast.LENGTH_SHORT).show();
                    }
                });

        BiometricPrompt.PromptInfo promptInfo = new BiometricPrompt.PromptInfo.Builder()
                .setTitle("Authorization")
                .setSubtitle("You must authenticate to continue: " + actionName)
                // Aligns with modern Android security standards (Strong Biometrics + PIN fallback)
                .setAllowedAuthenticators(BiometricManager.Authenticators.BIOMETRIC_STRONG | 
                                          BiometricManager.Authenticators.DEVICE_CREDENTIAL)
                .build();

        // Trigger the challenge
        biometricPrompt.authenticate(promptInfo);
    }
}

File: android/src/main/java/com/reveila/android/SovereignMemoryManager.kt
package com.reveila.android

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.widget.Toast
import androidx.biometric.BiometricManager
import androidx.biometric.BiometricPrompt
import androidx.core.content.ContextCompat
import androidx.fragment.app.FragmentActivity
import java.util.concurrent.Executor

/**
 * Ensures that high-risk Sovereign Memory writes or system actions
 * are gatekept behind a hardware-backed Biometric prompt.
 */
class SovereignMemoryManager(private val context: Context) {

    private val TAG = "SovereignMemoryManager"

    // [ ] 3. The Watchdog: Wrap SovereignMemoryManager writes/actions in BiometricPrompt challenge
    fun secureWriteOperation(activity: FragmentActivity, actionName: String, action: () -> Unit) {
        val executor: Executor = ContextCompat.getMainExecutor(activity)

        val biometricPrompt = BiometricPrompt(activity, executor,
            object : BiometricPrompt.AuthenticationCallback() {
                override fun onAuthenticationError(errorCode: Int, errString: CharSequence) {
                    super.onAuthenticationError(errorCode, errString)
                    Log.e(TAG, "Biometric authentication error for $actionName: $errString")
                    
                    // Show a Toast so the user actually sees why the prompt failed/aborted
                    Toast.makeText(activity, "Biometric Error: $errString\n(Did you set up a PIN/Fingerprint on this emulator?)", Toast.LENGTH_LONG).show()
                }

                override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
                    super.onAuthenticationSucceeded(result)
                    Log.i(TAG, "Biometric verified successfully. Executing $actionName...")
                    action() // Execute the secure action
                }

                override fun onAuthenticationFailed() {
                    super.onAuthenticationFailed()
                    Log.w(TAG, "Biometric authentication failed. Action $actionName blocked.")
                    Toast.makeText(activity, "Authentication failed. Try again.", Toast.LENGTH_SHORT).show()
                }
            })

        val promptInfo = BiometricPrompt.PromptInfo.Builder()
            .setTitle("Reveila High-Risk Action Guard")
            .setSubtitle("Authenticate to authorize: $actionName")
            // Allow fallback to standard PIN/Pattern if fingerprint isn't enrolled (crucial for Emulators)
            .setAllowedAuthenticators(BiometricManager.Authenticators.BIOMETRIC_STRONG or BiometricManager.Authenticators.DEVICE_CREDENTIAL)
            .build()

        // Trigger the challenge
        biometricPrompt.authenticate(promptInfo)
    }
}
